src/compiler/checker.ts

```diff
@@ -11985,6 +12025,12 @@ namespace ts {
             return !!(getObjectFlags(type) & ObjectFlags.Anonymous) && isEmptyObjectType(type);
         }
 
+        function isStringIndexSignatureOnlyType(type: Type): boolean {
+            return type.flags & TypeFlags.Object && getPropertiesOfType(type).length === 0 && getIndexInfoOfType(type, IndexKind.String) && !getIndexInfoOfType(type, IndexKind.Number) ||
+                type.flags & TypeFlags.UnionOrIntersection && every((<UnionOrIntersectionType>type).types, isStringIndexSignatureOnlyType) ||

```

Comment:
I want to exclude unions and intersections that contain anything but object types (e.g. instantiable types) which wouldn't happen if I just accessed props/indexes on the types directly. I really just care about pure string index signature types and unions/intersections thereof. Anything else can just go through the regular code path.

Reply:
Doesn't that mean the predicate you want here is
```ts
return (type.flags & TypeFlags.Object || (type.flags & TypeFlags.UnionOrIntersection && every((<UnionOrIntersectionType>type).types, t => !!(t.flags & TypeFlags.Object) || !!(t.flags & TypeFlags.Intersection && every((t as IntersectionType).types, t => !!(t.flags & TypeFlags.Intersection))))))
    && (getPropertiesOfType(type).length === 0 && getIndexInfoOfType(type, IndexKind.String) && !getIndexInfoOfType(type, IndexKind.Number))
```
?

Reply:
I'm not sure how that's better? There's no reason to check the properties and index signatures of unions and intersections themselves if we're already checking the constituents.

Reply:
Because, as I said, as written this fails to flag `{} & {[idx: string]: whatever}` as index-signature only?

Reply:
But we eliminate empty object types from intersections so that shouldn't occur.

Reply:
We don't though - we leave around _one_ empty object in an intersection type always (to keep around its implied-objectiness). We eliminate redundant ones. Usually (I've seen a couple cases where that fails to occur because we use `isEmptyAnonymousObjectType` for that check and so only remove redundant _anonymous_ empty objects!).

Reply:
If I'm not mistaken, we only keep around an empty object type if there are no other object types in the intersection. So, for `{} & { [key: string]: xxx }` we do indeed eliminate the `{}`.

Reply:
@ahejlsberg empty mapped types (eg, `{[K in never]: whatever}`) are for sure kept around - they're not _anonymous_ object types at all.

Reply:
That's fine, I don't care about optimizing strange combinations of empty mapped types and pure string index signature types. I really think the check is fine as is.